File: Xamarin/SampleXamarin.Android/SensorStatusView.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

using Android.Content;
using Android.Content.Res;
using Android.Util;
using Android.Widget;
using System;
using Android.Graphics;

namespace SampleXamarin
{
    public class SensorStatusView : LinearLayout
    {
        private TextView geoLocationStatusText;
        private TextView wifiStatusText;
        private TextView bluetoothStatusText;

        public SensorStatusView(Context context)
            : base(context)
        {
            Initialize(context);
        }

        public SensorStatusView(Context context, IAttributeSet attrs)
            : base(context, attrs)
        {
            Initialize(context);
        }

        public SensorStatusView(Context context, IAttributeSet attrs, int defStyleAttr)
            : base(context, attrs, defStyleAttr)
        {
            Initialize(context);
        }

        public SensorStatusModel Model { get; set; }

        public void Update()
        {
            geoLocationStatusText.SetTextColor(GetStatusColor(Model?.GeoLocationStatus));
            wifiStatusText.SetTextColor(GetStatusColor(Model?.WifiSignalStatus));
            bluetoothStatusText.SetTextColor(GetStatusColor(Model?.BluetoothSignalStatus));
        }

        private void Initialize(Context context)
        {
            Orientation = Android.Widget.Orientation.Vertical;
            Inflate(context, Resource.Layout.sensor_status, this);
            geoLocationStatusText = FindViewById<TextView>(Resource.Id.geolocation_status);
            wifiStatusText = FindViewById<TextView>(Resource.Id.wifi_status);
            bluetoothStatusText = FindViewById<TextView>(Resource.Id.bluetooth_status);
        }
        private Color GetStatusColor(SensorStatus? status)
        {
            if (!status.HasValue)
                return new Color(Context.GetColor(Resource.Color.sensorStatusIndeterminate));

            switch (status.Value)
            {
                case SensorStatus.Indeterminate:
                    return new Color(Context.GetColor(Resource.Color.sensorStatusIndeterminate));
                case SensorStatus.Blocked:
                    return new Color(Context.GetColor(Resource.Color.sensorStatusBlocked));
                case SensorStatus.Unavailable:
                    return new Color(Context.GetColor(Resource.Color.sensorStatusUnavailable));
                case SensorStatus.Available:
                    return new Color(Context.GetColor(Resource.Color.sensorStatusAvailable));
            }

            throw new ArgumentException("Invalid sensor status");
        }
    }
}

File: Android/Java/app/src/main/java/com/microsoft/sampleandroid/SensorStatusView.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
package com.microsoft.sampleandroid;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.widget.LinearLayout;
import android.widget.TextView;

public class SensorStatusView extends LinearLayout {
    private SensorStatusModel model;

    private TextView geoLocationStatusIcon;
    private TextView wifiStatusIcon;
    private TextView bluetoothStatusIcon;

    public SensorStatusView(Context context) {
        super(context);
        init(context);
    }

    public SensorStatusView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public SensorStatusView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    public void setModel(SensorStatusModel model) {
        this.model = model;
    }

    public void update() {
        geoLocationStatusIcon.setTextColor(getStatusColor(model.getGeoLocationStatus()));
        wifiStatusIcon.setTextColor(getStatusColor(model.getWifiSignalStatus()));
        bluetoothStatusIcon.setTextColor(getStatusColor(model.getBluetoothSignalStatus()));
    }

    private void init(Context context) {
        setOrientation(LinearLayout.VERTICAL);
        inflate(context, R.layout.sensor_status, this);
        geoLocationStatusIcon = findViewById(R.id.geolocation_status);
        wifiStatusIcon = findViewById(R.id.wifi_status);
        bluetoothStatusIcon = findViewById(R.id.bluetooth_status);
    }

    private int getStatusColor(SensorStatus status) {
        switch (status) {
            case Indeterminate:
                return ContextCompat.getColor(getContext(), R.color.sensorStatusIndeterminate);
            case Blocked:
                return ContextCompat.getColor(getContext(), R.color.sensorStatusBlocked);
            case Unavailable:
                return ContextCompat.getColor(getContext(), R.color.sensorStatusUnavailable);
            case Available:
                return ContextCompat.getColor(getContext(), R.color.sensorStatusAvailable);
        }

        throw new IllegalArgumentException("Invalid sensor status");
    }
}
